### PR TITLE
Merge adjacent StringIn nodes in oneOf

### PR DESCRIPTION
**Pull request by johnynek**
The main idea here is to merge adjacent Str and StringIn nodes in order to take advantage of the log vs linear complexity of matching that the RadixNode offers

cc @satabin

### CODE REVIEW

File: core/shared/src/main/scala/cats/parse/Parser.scala [hunk 1]
```diff
@@ -617,6 +617,18 @@ object Parser {
             Impl.rangesFor(ary).map { case (l, u) => InRange(o, l, u) }.toList
           }
           .toList
+      // merge the OneOfStr
+      val oosMerge: List[OneOfStr] =
```
Comment (johnynek) [inline core/shared/src/main/scala/cats/parse/Parser.scala:621]: this is needed to pass the laws once we are merging Str and StringIn nodes.

File: core/shared/src/main/scala/cats/parse/RadixNode.scala [hunk 2]
```diff
@@ -35,33 +36,34 @@ class RadixNode(
...
+    keys match {
+      case key :: keys =>
+        val prefixSize = commonPrefixLength(prefix, key)
+        if (prefixSize == 0) {
+          // no common prefix, group current suffixes together sorted again
+          groupByNonEmptyPrefix(
+            keys,
+            key,
+            NonEmptyList.one(key),
+            (prefix(0), prefix.tail, current.map(_.drop(prefix.size)).reverse) :: acc
```
Comment (johnynek) [inline core/shared/src/main/scala/cats/parse/RadixNode.scala:56]: I changed this to have the tail of the prefix so we don't compare twice the head when matching.

File: core/shared/src/main/scala/cats/parse/RadixNode.scala [hunk 3]
```diff
@@ -81,24 +86,21 @@ object RadixNode {
...
 
   private val leaf = new RadixNode(Array.empty, Array.empty, Array.empty, true)
 
-  private def commonPrefix(s1: String, s2: String): Int = {
-    @tailrec
-    def loop(idx: Int): Int =
-      if (idx >= s1.size || idx >= s2.size) {
-        idx
+  final def commonPrefixLength(s1: String, s2: String): Int = {
+    val len = Integer.min(s1.length, s2.length)
```
Comment (johnynek) [inline core/shared/src/main/scala/cats/parse/RadixNode.scala:93]: minor optimization to do only one comparison below, and do it against a local not the result of a method call.

**Comment by johnynek** [comment]
@regadas would you be able to review this one?

**Comment by regadas** [comment]
@johnynek yea sorry took a few days off 😄

**Review by regadas — APPROVED** [review]
LGTM

**Comment by johnynek** [comment]
Thanks for the review @regadas I have incorporated your suggestions.

**Pull request merged**